### ISSUE TITLE
SQS as Celery broker requires region configuration

### ISSUE THREAD

**Issue by villasv** [villasv/turbine#51]
Now that the stack targets any region, this configuration should be passed to Celery.

http://docs.celeryproject.org/en/latest/getting-started/brokers/sqs.html#region

Relevant section: https://github.com/apache/incubator-airflow/blob/master/airflow/config_templates/default_airflow.cfg#L416

**Comment by tomfaulhaber** [comment]
It turns out that there are a couple of problems here:

__Problem 1:__ 

Celery's kombu messaging layer doesn't ask boto if a region is configured by the user (as specified here: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#configuration). 

To do this, we need to add code to get the region from boto3 to `Channel.region` in transport/SQS.py. The code to get the region, if configured, is `boto3.session.Session().region_name` which will return the region as a string if it's set and `None` otherwise. (Method is here: https://github.com/celery/kombu/blob/master/kombu/transport/SQS.py#L469)

__Problem 2:__ 

If you specify the region in your airflow.cfg like this:

```
[celery_broker_transport_options]
region = us-east-2
```

it gets ignored. 

This is ignored because Airflow checks to see if the `visibility_timeout` parameter is set in the transport options and, if it's not set, replaces the transport options completely with `{'visibility_timeout': 21600}` (at https://github.com/apache/incubator-airflow/blob/master/airflow/config_templates/default_celery.py#L40).

If you add `visibility_timeout`, the region is correctly set:

```
[celery_broker_transport_options]
region = us-east-2
visibility_timeout = 21600
```

__Actions:__

There are four actions to take here:
1. Create an issue and PR for kombu to have them correctly get the region from the AWS config.
2. Create an issue and PR for Airflow to have them add `visibility_timeout` incrementally to the options rather than replace the options completely. We should do this quickly so that it can make the 1.10.1 release.
3. Add documentation to the ReadMe that shows how to correctly add the transport options to your airflow.cfg.
4. Modify the default airflow.cfg after it's built to include the correct region so that users can just change the executor to celery without having to do anything else.

1 & 2 will take a while to hit, but once 1 hits, we won't need to do any configuration in the airflow.cfg and then 3 & 4 can be removed.

WDYT, @villasv?

**Comment by villasv** [comment]
IIRC the release cycle for Celery can be very slow sometimes. I've waited months for them to release the merged fix for a dependency problem I've found (#14), so I don't think we should count on 1 soon enough. I think we could make the PR, but have a temporary fix on our side and remember to clear it later (like in #18). Airflow is currently on beta, so maybe PR for 2 would be merged quite fast. Anyway, that's a general bug that affects other things so I think we should make that PR anyway.

So I think we can go with:

1. Make the kombu PR
2. Make the Airflow PR
3. Try to solve it ourselves, but using environment variables instead. Probably defining `AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__{REGION,VISIBILITY_TIMEOUT}` (may the gods forgive me for this long name). I think this should work, because AFAIK everything in `airflow.cfg` can be overridden with env vars, so we can do that in the template directly.
4. Create an issue for removing the visibility var technical debt in case PR 2 is released
5. Create an issue for removing the region var technical debt in case PR 1 is released

This way there's no need for new instructions, it's still transparent to the users of the template.

**Comment by tomfaulhaber** [comment]
@villasv Oh that sounds good. I didn't realize that we could do those shenanigans with the environment variables. I'll do a test on that later today and, if it works, I'll do a PR.

From observation, that was my feeling about celery as well.

**Comment by tomfaulhaber** [comment]
Sadly, `AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__{REGION,VISIBILITY_TIMEOUT}` env vars don't work. This is because the broker transport options are loaded into a dict as a whole section, using `AirflowConfigParser.getsection()` rather than `AirflowConfigParser.get()` which is the function that looks for the environment variable override. (All in airflow/configuration.py)

**Comment by villasv** [comment]
:-( so the only way to fix this permanently (without making users adjust their `airflow.cfg`) is to merge that fix in kombu (latest release in May!) and then Celery...

Or patch Airflow to make it scan env vars with `AirflowConfigParser.getsection()` as well.
I'll get in touch with the fellows at Airflow about that config, meanwhile it seems that your original proposals 3 and/or 4 are the only way out.

**Comment by villasv** [comment]
I'll add a note about that in the README, but fixing the default airflow.cfg is going to be a whole new problem... specially because we don't have it in the filesystem during `cfn-init`.

Because the default config doesn't even use Celery, I think we can expect users to alter both the executor and the region in the transport options for now... even though I'm aware some users might miss that.

**Comment by villasv** [comment]
All right, I have green light to make a PR that extends `getsection()` to scan for environment variables as well, so we'll be able to work around that soon enough.

**Comment by villasv** [comment]
Now waiting on the next release of either Airflow (>=1.10.2) or Kombu (>=4.2.3)
(a Kombu release suffices because Celery will pick it up, it's requirement is `kombu>=4.2.0,<5.0`)

**Comment by villasv** [comment]
Airflow 1.10.2 has been released, so we can unblock this and solve it with environment variables.

**Comment by villasv** [comment]
Closed in 1ba9c338db5c552088b009b0a3b53752ff8858ea, finally
